### PR TITLE
Update the `app deploy` command for deployments release

### PR DESCRIPTION
**Pull request by mattnoakes**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/app-deploys/issues/429

### WHAT is this pull request doing?

It does a few things to get the `app deploy` command ready for deployments release based on the mocks in the issue description (sorry for the scope creep!):
- Replaces the `createDeployment` mutation with the `appDeploy` mutation
- Adds support for a `--no-release` flag when running `app deploy` which, when used, skips releasing the created app version
- Adds a warning or info banner (depending on the unified deployment beta flag) that describes the changes being made to the `app deploy` command when it is run
- Updates the copy in the `app deploy` prompts and output info depending on the presence of the `--no-release` flag
- Updates the copy in the success banners (which also depend on the `--no-release` flag)
- Adds support for a general errors section (mainly for the case when an error is returned for a version that can't be released because it requires review)

<details><summary>Running `app deploy` when the deployments beta is disabled</summary>

19-34-tmitz-avl0q

</details> 

<details><summary>Running `app deploy` when the deployments beta is enabled</summary>

##### `app deploy --no-release`
create-only

##### `app deploy`
create-and-release

##### `app deploy` when the deployment requires review
release-error

</details> 

### How to test your changes?

1. Create an app and generate an extension.
2. Run `app deploy` with the deployments beta disabled and verify that the extension is pushed to Shopify.
3. Enable the `app_unified_deployment` app beta flag and run `app deploy --no-release`. Verify that a deployment is created but **not** released.
4. Run `app deploy` and verify that a deployment can be both created **and** released.
5. Create an offsite payment provider extension (which requires approval) in the Partner dashboard; then, run `app deploy` and verify that a deployment is created but **not** released because of an error.

Note: Verify that the prompts, banners, and output info match the mocks while running through these flows.

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Review by matteodepalo — APPROVED** [review]
Nice one! 👍

File: packages/app/src/cli/services/deploy.ts
Comment (mattnoakes) [reply packages/app/src/cli/services/deploy.ts]: Thanks for the suggestion! I implemented it in [Use deploymentMode to represent deployment state](https://github.com/Shopify/cli/pull/1930/commits/46c6d542aa19fc93d80832a06012106a974db36a).

File: packages/app/src/cli/services/deploy/upload.ts
Comment (mattnoakes) [reply packages/app/src/cli/services/deploy/upload.ts]: Sounds good, thanks! I added this in [Handle output for multiple generic errors](https://github.com/Shopify/cli/pull/1930/commits/fbe862c30321a9dfd88c9221fac5897487037fc5).

File: packages/app/src/cli/services/context.ts
Comment (mattnoakes) [reply packages/app/src/cli/services/context.ts:311]: Good points!
> Also, if we ever remove the unifiedAppDeployment beta flag, this code would make the client go back to the 'legacy' version as undefined is falsey.

Could the function that calculates the `deploymentMode` value be changed to assume that an `undefined` value means that the beta flag has been removed (to prevent `deploymentMode` from falling back to the `legacy` value)? I.e.:
```
  const deploymentMode: DeploymentMode = (function () {
    if (partnersApp.betas?.unifiedAppDeployment || (typeof partnersApp.betas?.unifiedAppDeployment === "undefined")) {
      if (noRelease) {
        return 'unified-skip-release'
      } else {
        return 'unified'
      }
    } else {
      return 'legacy'
    }
  })()
```

> I heard that we'd prefer to always show a message. If so, wouldn't it be a bit redundant to show it to someone that has opted in and has know about the feature for a while?

I'll defer to @jessehoek /others, but how do other versioned packages like the CLI deal with temporary info messages?

**Comment by ani-shopify** [comment]
@matteodepalo I was chatting with @shauns about strategies in regards to merging this to main vs holding off for now until everything related to deployments that needs to be part of 3.48 is fully integrated into this branch.I'm leaning towards keeping this into a separate branch for now in order for these changes to be not included in 3.47. What do you think?

And if we go for the latter, will we be able to release a CLI version that could be used by someone like Eytan without setting up CLI locally? So far we've told people to use the nightly version. But would be great if there was a way to use a special CLI version that points to a branch. Is that possible?

**Comment by ani-shopify** [comment]
We talked about this during our meeting and have decided to not merge this PR into main. We will treat this branch as the branch that will hold all the features we want to release in 3.48. It will include the following tickets:

- https://github.com/Shopify/app-deploys/issues/474
- https://github.com/Shopify/app-deploys/issues/436
- https://github.com/Shopify/app-deploys/issues/437
- https://github.com/Shopify/app-deploys/issues/474
- https://github.com/Shopify/app-deploys/issues/471

P.S. Alternatively we can also create a new `deployment-3.48` branch and merge this feature branch into that branch.

**Comment by ani-shopify** [comment]
I just created `deployments-for-3-48` branch off of this branch. I'd recommend to close this PR for now. Thoughts or objections?

**Comment by mattnoakes** [comment]
@matteodepalo that's kinda my understanding. For example, the urls in the warning/info/success banners aren't ready. We also want to show [a different info banner](https://www.figma.com/file/dilbrKb0vthWTErdYgUunQ/Unified-App-Deployment?type=design&node-id=7469-45465&t=iWL1TqpW6HGpkkX7-0) in 3.47, then show [the one in this PR](https://www.figma.com/file/dilbrKb0vthWTErdYgUunQ/Unified-App-Deployment?type=design&node-id=7469-45519&t=3QXDR5k05vO6vuYh-0) in 3.48. Plus, we haven't figured out how the opt-into-deployments flow will work yet (which might change the content of the info banner in this PR).
> The only problem is, if we don't release this branch we'll have the old mutation being used when 3.47 goes out.

I don't see a problem with this. What about you, @ani-shopify?

**Comment by ani-shopify** [comment]
> I don't see a problem with this. What about you, @ani-shopify?

Yes, I we are ok with that.

**Pull request closed without merging**
